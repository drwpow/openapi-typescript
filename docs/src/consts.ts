export const SITE = {
  title: "OpenAPI TS",
  description: "Tools for consuming OpenAPI schemas in TypeScript.",
  defaultLanguage: "en-us",
} as const;

export const OPEN_GRAPH = {
  image: {
    src: "https://github.com/drwpow/openapi-typescript/blob/main/.github/assets/banner.png?raw=true",
  },
};

export const GITHUB_EDIT_URL = `https://github.com/drwpow/openapi-typescript/tree/main/docs`;

export const KNOWN_LANGUAGES = {
  English: "en",
} as const;
export const KNOWN_LANGUAGE_CODES = Object.values(KNOWN_LANGUAGES);
