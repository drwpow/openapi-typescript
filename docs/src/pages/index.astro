---
import MainLayout from "../layouts/MainLayout.astro";
---

<MainLayout title="openapi-typescript">
  <p>Consume <a href="https://spec.openapis.org/oas/latest.html" target="_blank" rel="noopener noreferrer">OpenAPI 3.0 & 3.1 schemas</a> in TypeScript in the lightest, fastest way possible.</p>
  <div class="selection">
    <a class="lib lib--ts" href="/introduction">
      <img src="/assets/openapi-ts.svg" />
      <span>Convert OpenAPI schemas to TypeScript</span>
    </a>
    <a class="lib lib--fetch" href="/openapi-fetch">
      <img src="/assets/openapi-fetch.svg" />
      <span>Fast, typesafe fetch client for your OpenAPI schema</span>
    </a>
  </div>
</MainLayout>

<style lang="scss">
  @use '../tokens' as *;

  .selection {
    display: grid;
    gap: 2rem;
    grid-template-columns: repeat(auto-fit, minmax(240px, 1fr));
  }

  .lib {
    align-items: flex-start;
    border-radius: 1rem;
    border: 2px solid token('color.ui.contrast.10');
    color: inherit;
    display: flex;
    flex-direction: column;
    font-size: 1.125rem;
    gap: 0.5rem;
    line-height: 1.5;
    padding: 1rem;
    position: relative;
    transition: background-color 50ms linear, border-color 50ms linear;
    text-decoration: none;

    &:hover {
      text-decoration: none;
    }

    img {
      width: 12rem;
      height: auto;
    }

    &--ts {
      &:hover,
      &:focus-visible {
        background-color: color-mix(in oklab, #{token('color.brand.ts-blue')}, 95% transparent);
        border-color: token('color.brand.ts-blue');
      }

    }

    &--fetch {
      &:hover,
      &:focus-visible {
        background-color: color-mix(in oklab, #{token('color.brand.fetch-green')}, 95% transparent);
        border-color: token('color.brand.fetch-green');
      }
    }
  }
</style>
