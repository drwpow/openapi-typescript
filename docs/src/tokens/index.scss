////
/// Design Tokens
/// Autogenerated from tokens.json.
/// DO NOT EDIT!
////

@use "sass:list";
@use "sass:map";

$__token-values: (
  "color.blue.10": (
    default: (var(--color-blue-10)),
  ),
  "color.blue.15": (
    default: (var(--color-blue-15)),
  ),
  "color.blue.20": (
    default: (var(--color-blue-20)),
  ),
  "color.blue.25": (
    default: (var(--color-blue-25)),
  ),
  "color.blue.30": (
    default: (var(--color-blue-30)),
  ),
  "color.blue.40": (
    default: (var(--color-blue-40)),
  ),
  "color.blue.50": (
    default: (var(--color-blue-50)),
  ),
  "color.blue.60": (
    default: (var(--color-blue-60)),
  ),
  "color.blue.70": (
    default: (var(--color-blue-70)),
  ),
  "color.blue.80": (
    default: (var(--color-blue-80)),
  ),
  "color.blue.85": (
    default: (var(--color-blue-85)),
  ),
  "color.blue.90": (
    default: (var(--color-blue-90)),
  ),
  "color.blue.95": (
    default: (var(--color-blue-95)),
  ),
  "color.blue.100": (
    default: (var(--color-blue-100)),
  ),
  "color.brand.fetch-green": (
    default: (var(--color-brand-fetch-green)),
  ),
  "color.brand.ts-blue": (
    default: (var(--color-brand-ts-blue)),
  ),
  "color.gray.10": (
    default: (var(--color-gray-10)),
  ),
  "color.gray.15": (
    default: (var(--color-gray-15)),
  ),
  "color.gray.20": (
    default: (var(--color-gray-20)),
  ),
  "color.gray.25": (
    default: (var(--color-gray-25)),
  ),
  "color.gray.30": (
    default: (var(--color-gray-30)),
  ),
  "color.gray.40": (
    default: (var(--color-gray-40)),
  ),
  "color.gray.50": (
    default: (var(--color-gray-50)),
  ),
  "color.gray.60": (
    default: (var(--color-gray-60)),
  ),
  "color.gray.70": (
    default: (var(--color-gray-70)),
  ),
  "color.gray.80": (
    default: (var(--color-gray-80)),
  ),
  "color.gray.85": (
    default: (var(--color-gray-85)),
  ),
  "color.gray.90": (
    default: (var(--color-gray-90)),
  ),
  "color.gray.95": (
    default: (var(--color-gray-95)),
  ),
  "color.gray.100": (
    default: (var(--color-gray-100)),
  ),
  "color.ui.action": (
    default: (var(--color-ui-action)),
  ),
  "color.ui.bg": (
    default: (var(--color-ui-bg)),
    "light": (var(--color-ui-bg)),
    "dark": (var(--color-ui-bg)),
  ),
  "color.ui.bg-offset": (
    default: (var(--color-ui-bg-offset)),
    "light": (var(--color-ui-bg-offset)),
    "dark": (var(--color-ui-bg-offset)),
  ),
  "color.ui.border": (
    default: (var(--color-ui-border)),
    "light": (var(--color-ui-border)),
    "dark": (var(--color-ui-border)),
  ),
  "color.ui.callout-bg": (
    default: (var(--color-ui-callout-bg)),
    "light": (var(--color-ui-callout-bg)),
    "dark": (var(--color-ui-callout-bg)),
  ),
  "color.ui.contrast.00": (
    default: (var(--color-ui-contrast-00)),
    "light": (var(--color-ui-contrast-00)),
    "dark": (var(--color-ui-contrast-00)),
  ),
  "color.ui.contrast.05": (
    default: (var(--color-ui-contrast-05)),
    "light": (var(--color-ui-contrast-05)),
    "dark": (var(--color-ui-contrast-05)),
  ),
  "color.ui.contrast.10": (
    default: (var(--color-ui-contrast-10)),
    "light": (var(--color-ui-contrast-10)),
    "dark": (var(--color-ui-contrast-10)),
  ),
  "color.ui.contrast.15": (
    default: (var(--color-ui-contrast-15)),
    "light": (var(--color-ui-contrast-15)),
    "dark": (var(--color-ui-contrast-15)),
  ),
  "color.ui.contrast.20": (
    default: (var(--color-ui-contrast-20)),
    "light": (var(--color-ui-contrast-20)),
    "dark": (var(--color-ui-contrast-20)),
  ),
  "color.ui.contrast.30": (
    default: (var(--color-ui-contrast-30)),
    "light": (var(--color-ui-contrast-30)),
    "dark": (var(--color-ui-contrast-30)),
  ),
  "color.ui.contrast.40": (
    default: (var(--color-ui-contrast-40)),
    "light": (var(--color-ui-contrast-40)),
    "dark": (var(--color-ui-contrast-40)),
  ),
  "color.ui.contrast.50": (
    default: (var(--color-ui-contrast-50)),
    "light": (var(--color-ui-contrast-50)),
    "dark": (var(--color-ui-contrast-50)),
  ),
  "color.ui.contrast.60": (
    default: (var(--color-ui-contrast-60)),
    "light": (var(--color-ui-contrast-60)),
    "dark": (var(--color-ui-contrast-60)),
  ),
  "color.ui.contrast.70": (
    default: (var(--color-ui-contrast-70)),
    "light": (var(--color-ui-contrast-70)),
    "dark": (var(--color-ui-contrast-70)),
  ),
  "color.ui.contrast.75": (
    default: (var(--color-ui-contrast-75)),
    "light": (var(--color-ui-contrast-75)),
    "dark": (var(--color-ui-contrast-75)),
  ),
  "color.ui.contrast.80": (
    default: (var(--color-ui-contrast-80)),
    "light": (var(--color-ui-contrast-80)),
    "dark": (var(--color-ui-contrast-80)),
  ),
  "color.ui.contrast.85": (
    default: (var(--color-ui-contrast-85)),
    "light": (var(--color-ui-contrast-85)),
    "dark": (var(--color-ui-contrast-85)),
  ),
  "color.ui.contrast.90": (
    default: (var(--color-ui-contrast-90)),
    "light": (var(--color-ui-contrast-90)),
    "dark": (var(--color-ui-contrast-90)),
  ),
  "color.ui.contrast.95": (
    default: (var(--color-ui-contrast-95)),
    "light": (var(--color-ui-contrast-95)),
    "dark": (var(--color-ui-contrast-95)),
  ),
  "color.ui.contrast.100": (
    default: (var(--color-ui-contrast-100)),
    "light": (var(--color-ui-contrast-100)),
    "dark": (var(--color-ui-contrast-100)),
  ),
  "color.ui.fg": (
    default: (var(--color-ui-fg)),
    "light": (var(--color-ui-fg)),
    "dark": (var(--color-ui-fg)),
  ),
  "color.ui.text-subdue": (
    default: (var(--color-ui-text-subdue)),
    "light": (var(--color-ui-text-subdue)),
    "dark": (var(--color-ui-text-subdue)),
  ),
  "typography.base": (
    "__cobalt-error": "This is a typography mixin. Use `@include typography(\"typography.base\")` instead.",
  ),
  "typography.family.base": (
    default: (var(--typography-family-base)),
  ),
  "typography.family.mono": (
    default: (var(--typography-family-mono)),
  ),
  "typography.h1": (
    "__cobalt-error": "This is a typography mixin. Use `@include typography(\"typography.h1\")` instead.",
  ),
  "typography.h2": (
    "__cobalt-error": "This is a typography mixin. Use `@include typography(\"typography.h2\")` instead.",
  ),
  "typography.h3": (
    "__cobalt-error": "This is a typography mixin. Use `@include typography(\"typography.h3\")` instead.",
  ),
  "typography.h4": (
    "__cobalt-error": "This is a typography mixin. Use `@include typography(\"typography.h4\")` instead.",
  ),
  "typography.mono": (
    "__cobalt-error": "This is a typography mixin. Use `@include typography(\"typography.mono\")` instead.",
  ),
);

$__token-typography-mixins: (
  "typography.base": (
    default: (
      "font-family": (var(--typography-base-font-family)),
      "font-size": (var(--typography-base-font-size)),
      "font-weight": (var(--typography-base-font-weight)),
      "letter-spacing": (var(--typography-base-letter-spacing)),
      "line-height": (var(--typography-base-line-height)),
    ),
  ),
  "typography.h1": (
    default: (
      "font-family": (var(--typography-h1-font-family)),
      "font-size": (var(--typography-h1-font-size)),
      "font-weight": (var(--typography-h1-font-weight)),
      "letter-spacing": (var(--typography-h1-letter-spacing)),
      "line-height": (var(--typography-h1-line-height)),
    ),
  ),
  "typography.h2": (
    default: (
      "font-family": (var(--typography-h2-font-family)),
      "font-size": (var(--typography-h2-font-size)),
      "font-weight": (var(--typography-h2-font-weight)),
      "letter-spacing": (var(--typography-h2-letter-spacing)),
      "line-height": (var(--typography-h2-line-height)),
    ),
  ),
  "typography.h3": (
    default: (
      "font-family": (var(--typography-h3-font-family)),
      "font-size": (var(--typography-h3-font-size)),
      "font-weight": (var(--typography-h3-font-weight)),
      "letter-spacing": (var(--typography-h3-letter-spacing)),
      "line-height": (var(--typography-h3-line-height)),
    ),
  ),
  "typography.h4": (
    default: (
      "font-family": (var(--typography-h4-font-family)),
      "font-size": (var(--typography-h4-font-size)),
      "font-weight": (var(--typography-h4-font-weight)),
      "letter-spacing": (var(--typography-h4-letter-spacing)),
      "line-height": (var(--typography-h4-line-height)),
    ),
  ),
  "typography.mono": (
    default: (
      "font-family": (var(--typography-mono-font-family)),
      "font-size": (var(--typography-mono-font-size)),
      "font-weight": (var(--typography-mono-font-weight)),
      "letter-spacing": (var(--typography-mono-letter-spacing)),
      "line-height": (var(--typography-mono-line-height)),
    ),
  ),
);

@function token($tokenName, $modeName: default) {
  @if map.has-key($__token-values, $tokenName) == false {
    @error "No token named \"#{$tokenName}\"";
  }
  $_token: map.get($__token-values, $tokenName);
  @if map.has-key($_token, "__cobalt-error") {
    @error map.get($_token, "__cobalt-error");
  }
  @if map.has-key($_token, $modeName) {
    @return map.get($_token, $modeName);
  } @else {
    @return map.get($_token, default);
  }
}

@function listModes($tokenName) {
  @if map.has-key($__token-values, $tokenName) == false {
    @error "No token named \"#{$tokenName}\"";
  }
  $_modes: ();
  @each $k in map.get($__token-values, $tokenName) {
    @if $k != "default" {
      $_modes: list.append($_modes, $k);
    }
  }
  @return $_modes;
}

@mixin typography($tokenName, $modeName: default) {
  @if map.has-key($__token-typography-mixins, $tokenName) == false {
    @error "No typography mixin named \"#{$tokenName}\"";
  }
  $_mixin: map.get($__token-typography-mixins, $tokenName);
  $_properties: map.get($_mixin, default);
  @if map.has-key($_mixin, $modeName) {
    $_properties: map.get($_mixin, $modeName);
  }
  @each $_property, $_value in $_properties {
    #{$_property}: #{$_value};
  }
}
