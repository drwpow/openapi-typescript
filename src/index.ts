import path from "path";
import prettier from "prettier";
import { swaggerVersion } from "./utils";
import {
  OpenAPI2,
  OpenAPI2Schemas,
  OpenAPI3,
  OpenAPI3Schemas,
  SwaggerToTSOptions,
} from "./types";
import v2 from "./v2";
import v3 from "./v3";

export const WARNING_MESSAGE = `/**
* This file was auto-generated by swagger-to-ts.
* Do not make direct changes to the file.
*/


`;

export default function swaggerToTS(
  schema: OpenAPI2 | OpenAPI2Schemas | OpenAPI3 | OpenAPI3Schemas,
  options?: SwaggerToTSOptions
): string {
  // generate types for V2 and V3
  const version =
    (options && options.version) ||
    swaggerVersion(schema as OpenAPI2 | OpenAPI3);
  let output = `${WARNING_MESSAGE}`;
  switch (version) {
    case 2: {
      output = output.concat(v2(schema as OpenAPI2 | OpenAPI2Schemas, options));
      break;
    }
    case 3: {
      output = output.concat(v3(schema as OpenAPI3 | OpenAPI3Schemas, options));
      break;
    }
  }

  // Prettify output
  let prettierOptions: prettier.Options = { parser: "typescript" };
  if (options && options.prettierConfig) {
    try {
      const userOptions = prettier.resolveConfig.sync(
        path.resolve(process.cwd(), options.prettierConfig)
      );
      prettierOptions = {
        ...prettierOptions,
        ...userOptions,
      };
    } catch (err) {
      console.error(`‚ùå ${err}`);
    }
  }
  return prettier.format(output, prettierOptions);
}
