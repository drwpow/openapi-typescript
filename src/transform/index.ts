import { comment } from "../utils";
import { transformHeaderObjMap } from "./headers";
import { transformOperationObj } from "./operation";
import { transformPathsObj } from "./paths";
import { transformResponsesObj, transformRequestBodies } from "./responses";
import { transformSchemaObjMap } from "./schema";

interface TransformOptions {
  rawSchema?: boolean;
  version: number;
}

export function transformAll(schema: any, { version, rawSchema }: TransformOptions): string {
  let output = "";

  let operations: Record<string, OperationObject> = {};

  // --raw-schema mode
  if (rawSchema) {
    switch (version) {
      case 2: {
        return `export interface definitions {\n  ${transformSchemaObjMap(schema, {
          required: Object.keys(schema),
        })}\n}`;
      }
      case 3: {
        return `export interface schemas {\n    ${transformSchemaObjMap(schema, {
          required: Object.keys(schema),
        })}\n  }\n\n`;
      }
    }
  }

  // #/paths (V2 & V3)
  output += `export interface paths {\n`; // open paths
  if (schema.paths) {
    output += transformPathsObj(
      schema.paths,
      {
        operations,
        parameters: (schema.components && schema.components.parameters) || schema.parameters,
      },
      version
    );
  }
  output += `}\n\n`; // close paths

  switch (version) {
    case 2: {
      // #/definitions
      if (schema.definitions) {
        output += `export interface definitions {\n  ${transformSchemaObjMap(schema.definitions, {
          required: Object.keys(schema.definitions),
        })}\n}\n\n`;
      }

      // #/parameters
      if (schema.parameters) {
        const required = Object.keys(schema.parameters);
        output += `export interface parameters {\n    ${transformSchemaObjMap(schema.parameters, {
          required,
        })}\n  }\n\n`;
      }

      // #/parameters
      if (schema.responses) {
        output += `export interface responses {\n    ${transformResponsesObj(schema.responses)}\n  }\n\n`;
      }
      break;
    }
    case 3: {
      // #/components
      output += `export interface components {\n`; // open components

      if (schema.components) {
        // #/components/schemas
        if (schema.components.schemas) {
          const required = Object.keys(schema.components.schemas);
          output += `  schemas: {\n    ${transformSchemaObjMap(schema.components.schemas, { required })}\n  }\n`;
        }

        // #/components/responses
        if (schema.components.responses) {
          output += `  responses: {\n    ${transformResponsesObj(schema.components.responses)}\n  }\n`;
        }

        // #/components/parameters
        if (schema.components.parameters) {
          const required = Object.keys(schema.components.parameters);
          output += `  parameters: {\n    ${transformSchemaObjMap(schema.components.parameters, {
            required,
          })}\n  }\n`;
        }

        // #/components/requestBodies
        if (schema.components.requestBodies) {
          output += `  requestBodies: {\n    ${transformRequestBodies(schema.components.requestBodies)}\n  }\n`;
        }

        // #/components/headers
        if (schema.components.headers) {
          output += `  headers: {\n    ${transformHeaderObjMap(schema.components.headers)}  }\n`;
        }
      }

      output += `}\n\n`; // close components
      break;
    }
  }

  output += `export interface operations {\n`; // open operations
  if (Object.keys(operations).length) {
    Object.entries(operations).forEach(([operationId, operation]) => {
      if (operation.description) output += comment(operation.description); // handle comment
      output += `  "${operationId}": {\n    ${transformOperationObj(
        operation,
        version,
        schema.components && schema.components.parameters
      )}\n  }\n`;
    });
  }
  output += `}\n`; // close operations

  return output.trim();
}
