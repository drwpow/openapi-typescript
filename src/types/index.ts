import { OpenAPI2SchemaObject } from "./OpenAPI2";
import { OpenAPI3SchemaObject } from "./OpenAPI3";

export * from "./OpenAPI2";
export * from "./OpenAPI3";

export interface Property {
  interfaceType: string;
  optional: boolean;
  description?: string;
}

export interface SwaggerToTSOptions {
  /** (optional) Path to Prettier config */
  prettierConfig?: string;
  /** (optional) Function to iterate over every schema object before transforming to TypeScript */
  propertyMapper?: (
    schemaObject: OpenAPI2SchemaObject | OpenAPI3SchemaObject,
    property: Property
  ) => Property;
  /** (optional) Parsing input document as raw schema rather than OpenAPI document */
  rawSchema?: boolean;
  /** (optional) OpenAPI version. Must be present if parsing raw schema */
  version?: number;
}
