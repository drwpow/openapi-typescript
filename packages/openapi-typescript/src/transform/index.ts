import ts, { TypeAliasDeclaration, TypeLiteralNode } from "typescript";
import { NEVER, STRING, tsModifiers, tsRecord } from "../lib/ts.js";
import { createRef } from "../lib/utils.js";
import { GlobalContext, OpenAPI3 } from "../types.js";
import transformComponentsObject from "./components-object.js";
import transformPathsObject from "./paths-object.js";
import transformSchemaObject from "./schema-object.js";
import transformWebhooksObject from "./webhooks-object.js";

type SchemaTransforms = keyof Pick<
  OpenAPI3,
  "paths" | "webhooks" | "components" | "$defs"
>;

const transformers: Record<
  SchemaTransforms,
  (node: any, options: GlobalContext) => ts.TypeNode // eslint-disable-line @typescript-eslint/no-explicit-any
> = {
  paths: transformPathsObject,
  webhooks: transformWebhooksObject,
  components: transformComponentsObject,
  $defs: (node, options) =>
    transformSchemaObject(node, { path: createRef(["$defs"]), ctx: options }),
};

export default function transformSchema(schema: OpenAPI3, ctx: GlobalContext) {
  const type: ts.Node[] = [];

  for (const root of Object.keys(transformers) as SchemaTransforms[]) {
    if (schema[root]) {
      const subType = transformers[root](schema[root], ctx);
      type.push(
        ctx.exportType
          ? ts.factory.createTypeAliasDeclaration(
              /* modifiers      */ tsModifiers({
                export: true,
                readonly: ctx.immutableTypes,
              }),
              /* name           */ root,
              /* typeParameters */ undefined,
              /* type           */ subType,
            )
          : ts.factory.createInterfaceDeclaration(
              /* modifiers       */ tsModifiers({
                export: true,
                readonly: ctx.immutableTypes,
              }),
              /* name            */ root,
              /* typeParameters  */ undefined,
              /* heritageClauses */ undefined,
              /* members         */ (subType as TypeLiteralNode).members,
            ),
      );
    } else {
      type.push(
        ts.factory.createTypeAliasDeclaration(
          /* modifiers      */ tsModifiers({
            export: true,
            readonly: ctx.immutableTypes,
          }),
          /* name           */ root,
          /* typeParameters */ undefined,
          /* type           */ tsRecord(STRING, NEVER),
        ),
      );
    }
  }

  // inject
  let hasOperations = false;
  for (const injectedType of ctx.injectFooter) {
    if (
      (injectedType as TypeAliasDeclaration).name.escapedText === "operations"
    ) {
      hasOperations = true;
    }
    type.push(injectedType);
  }
  if (!hasOperations) {
    // if no operations created, inject empty operations type
    type.push(
      ts.factory.createTypeAliasDeclaration(
        /* modifiers      */ tsModifiers({
          export: true,
          readonly: ctx.immutableTypes,
        }),
        /* name           */ "operations",
        /* typeParameters */ undefined,
        /* type           */ tsRecord(STRING, NEVER),
      ),
    );
  }

  return type;
}
