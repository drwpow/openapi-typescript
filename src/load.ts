import fs from "fs";
import path from "path";
import { URL } from "url";
import fetch, { Headers } from "node-fetch";
import slash from "slash";
import mime from "mime";
import yaml from "js-yaml";
import { GlobalContext } from "./types";
import { parseRef } from "./utils";

type PartialSchema = Record<string, any>; // not a very accurate type, but this is easier to deal with before we know we’re dealing with a valid spec

function parseSchema(schema: any, type: "YAML" | "JSON") {
  if (type === "YAML") {
    try {
      return yaml.load(schema);
    } catch (err) {
      throw new Error(`YAML: ${err.toString()}`);
    }
  } else {
    try {
      return JSON.parse(schema);
    } catch (err) {
      throw new Error(`JSON: ${err.toString()}`);
    }
  }
}

function isFile(url: URL): boolean {
  return url.protocol === "file:";
}

export function resolveSchema(url: string): URL {
  // option 1: remote
  if (url.startsWith("http://") || url.startsWith("https://")) {
    return new URL(url);
  }

  // option 2: local
  const localPath = path.isAbsolute(url)
    ? new URL("", `file://${slash(url)}`)
    : new URL(url, `file://${slash(process.cwd())}/`); // if absolute path is provided use that; otherwise search cwd\
  if (!fs.existsSync(localPath)) {
    throw new Error(`Could not locate ${url}`);
  } else if (fs.statSync(localPath).isDirectory()) {
    throw new Error(`${localPath} is a directory not a file`);
  }
  return localPath;
}

interface LoadOptions extends GlobalContext {
  rootURL: URL;
  schemas: { [url: string]: PartialSchema };
}

// temporary cache for load()
let urlCache = new Set<string>(); // URL cache (prevent URLs from being loaded over and over)

/** Load a schema from local path or remote URL */
export default async function load(schemaURL: URL, options: LoadOptions): Promise<{ [url: string]: PartialSchema }> {
  if (urlCache.has(schemaURL.href)) return options.schemas; // exit early if this has already been scanned
  urlCache.add(schemaURL.href); // add URL to cache

  const schemas = options.schemas;

  let contents = "";
  let contentType = "";

  if (isFile(schemaURL)) {
    // load local
    contents = await fs.promises.readFile(schemaURL, "utf8");
    contentType = mime.getType(schemaURL.href) || "";
  } else {
    // load remote
    const headers = new Headers();
    headers.set("User-Agent", "openapi-typescript");
    if (options.auth) headers.set("Authorization", options.auth);
    const res = await fetch(schemaURL.href, { method: "GET", headers });
    contentType = res.headers.get("Content-Type") || "";
    contents = await res.text();
  }

  const isYAML = contentType === "application/openapi+yaml" || contentType === "text/yaml";
  const isJSON =
    contentType === "application/json" ||
    contentType === "application/json5" ||
    contentType === "application/openapi+json";
  if (isYAML) {
    schemas[schemaURL.href] = parseSchema(contents, "YAML");
  } else if (isJSON) {
    schemas[schemaURL.href] = parseSchema(contents, "JSON");
  } else {
    // if contentType is unknown, guess
    try {
      schemas[schemaURL.href] = parseSchema(contents, "JSON");
    } catch (err1) {
      try {
        schemas[schemaURL.href] = parseSchema(contents, "YAML");
      } catch (err2) {
        throw new Error(`Unknown format${contentType ? `: "${contentType}"` : ""}. Only YAML or JSON supported.`); // give up: unknown type
      }
    }
  }

  // scan $refs, but don’t transform (load everything in parallel)
  const refPromises: Promise<any>[] = [];
  schemas[schemaURL.href] = JSON.parse(JSON.stringify(schemas[schemaURL.href]), (k, v) => {
    if (k !== "$ref" || typeof v !== "string") return v;

    const { url: refURL } = parseRef(v);
    if (refURL) {
      // load $refs (only if new) and merge subschemas with top-level schema
      const nextURL =
        refURL.startsWith("http://") || refURL.startsWith("https://")
          ? new URL(refURL)
          : new URL(slash(refURL), schemaURL);
      refPromises.push(
        load(nextURL, options).then((subschemas) => {
          for (const subschemaURL of Object.keys(subschemas)) {
            schemas[subschemaURL] = subschemas[subschemaURL];
          }
        })
      );
      return v.replace(refURL, nextURL.href); // resolve relative URLs to absolute URLs so the schema can be flattened
    }
    return v;
  });
  await Promise.all(refPromises);

  // transform $refs once, at the root schema, after all have been scanned & downloaded (much easier to do here when we have the context)
  if (schemaURL.href === options.rootURL.href) {
    for (const subschemaURL of Object.keys(schemas)) {
      // transform $refs in schema
      schemas[subschemaURL] = JSON.parse(JSON.stringify(schemas[subschemaURL]), (k, v) => {
        if (k !== "$ref" || typeof v !== "string") return v;
        if (!v.includes("#")) return v; // already transformed; skip

        const { url, parts } = parseRef(v);
        // scenario 1: resolve all external URLs so long as they don’t point back to root schema
        if (url && new URL(url).href !== options.rootURL.href) {
          const relativeURL =
            isFile(new URL(url)) && isFile(options.rootURL)
              ? path.posix.relative(path.posix.dirname(options.rootURL.href), url)
              : url;
          return `external["${relativeURL}"]["${parts.join('"]["')}"]`; // export external ref
        }
        // scenario 2: treat all $refs in external schemas as external
        if (!url && subschemaURL !== options.rootURL.href) {
          const relativeURL =
            isFile(new URL(subschemaURL)) && isFile(options.rootURL)
              ? path.posix.relative(path.posix.dirname(options.rootURL.href), subschemaURL)
              : subschemaURL;
          return `external["${relativeURL}"]["${parts.join('"]["')}"]`; // export external ref
        }

        // scenario 3: transform all $refs pointing back to root schema
        const [base, ...rest] = parts;
        return `${base}["${rest.join('"]["')}"]`; // transform other $refs to the root schema (including external refs that point back to the root schema)
      });

      // use relative keys for external schemas (schemas generated on different machines should have the same namespace)
      if (subschemaURL !== options.rootURL.href) {
        const relativeURL =
          isFile(new URL(subschemaURL)) && isFile(options.rootURL)
            ? path.posix.relative(path.posix.dirname(options.rootURL.href), subschemaURL)
            : subschemaURL;
        if (relativeURL !== subschemaURL) {
          schemas[relativeURL] = schemas[subschemaURL];
          delete schemas[subschemaURL];
        }
      }
    }
  }

  return schemas;
}
