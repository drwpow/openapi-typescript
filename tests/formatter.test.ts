import { default as openapiTS } from "../src/index";

describe("formatter", () => {
  it("basic", async () => {
    const schema = {
      openapi: "3.0.1",
      components: {
        schemas: {
          date: {
            type: "string",
            format: "date-time",
          },
        },
      },
    };

    // Ensures consistency and indentation in string output with snapshot. Please do NOT update snapshot unless value changes
    expect(
      await openapiTS(schema, {
        formatter(schemaObj) {
          if (schemaObj.format === "date-time") {
            return "Date";
          }
          return;
        },
        version: 3,
      })
    ).toMatchSnapshot();
  });

  it("hasObject", async () => {
    const schemaHasObject = {
      openapi: "3.0.1",
      components: {
        schemas: {
          calendar: {
            type: "object",
            properties: {
              dates: {
                type: "object",
                properties: {
                  start: {
                    type: "string",
                    format: "date-time",
                  },
                },
              },
            },
          },
        },
      },
    };

    expect(
      await openapiTS(schemaHasObject, {
        formatter(schemaObj) {
          if (schemaObj.format === "date-time") {
            return "Date";
          }
          return;
        },
        version: 3,
      })
    ).toMatchSnapshot();
  });
});
