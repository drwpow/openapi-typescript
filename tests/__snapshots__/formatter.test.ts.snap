// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatter basic 1`] = `
"/**
 * This file was auto-generated by openapi-typescript.
 * Do not make direct changes to the file.
 */

export interface paths {}

export interface components {
  schemas: {
    date: Date;
  };
}

export interface operations {}

export interface external {}
"
`;

exports[`formatter hasObject 1`] = `
"/**
 * This file was auto-generated by openapi-typescript.
 * Do not make direct changes to the file.
 */

export interface paths {}

export interface components {
  schemas: {
    calendar: {
      dates?: {
        start?: Date;
      };
    };
  };
}

export interface operations {}

export interface external {}
"
`;
